### PR TITLE
Impl/geometric mean

### LINKED ISSUE #2019: Add GeometricMean
**Describe the feature and the current behavior/state.**

**Relevant information**
- Are you willing to contribute it (yes/no): YES
- Are you willing to maintain it going forward? (yes/no): YES
- Is there a relevant academic paper? (if so, where): NO
- Is there already an implementation in another framework? (if so, where): Pytorch Ignite Metrics
- Was it part of tf.contrib? (if so, where): NO

**Which API type would this fall under (layer, metric, optimizer, etc.)**
Metric

**Who will benefit with this feature?**
Researchers or Engineers who need a GeometricMean

**Any other info.**
None

### PR DESCRIPTION
**Pull request by marload**
# Description
I implemented `tfa.metrics.GeometricMean`

If you let me know what needs to be corrected, I will fix it.
Thank you!

Fixes #2019 

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation (Requires creating an issue first)
    - [ ] The changes conform to the [contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback (Requires creating an issue first)
    - [ ] The changes conform to the [contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition (Requires creating an issue first)
    - [ ] The changes conform to the [contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer (Requires creating an issue first)
    - [ ] The changes conform to the [contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss (Requires creating an issue first)
    - [ ] The changes conform to the [contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [x] New Metric (Requires creating an issue first)
    - [x] The changes conform to the [contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer (Requires creating an issue first)
    - [ ] The changes conform to the [contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell (Requires creating an issue first)
    - [ ] The changes conform to the [contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition (Requires creating an issue first)
    - [ ] The changes conform to the [contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition (Requires creating an issue first)
    - [ ] The changes conform to the [contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# How Has This Been Tested?

If you're adding a bugfix or new feature please describe the tests that you ran to verify your changes:
*  

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [x] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

### CODE REVIEW

File: tensorflow_addons/metrics/geometric_mean.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            warnings.warn(
+                "`sample_weight` is not None. Be aware that GeometricMean"
+                "does not take `sample_weight` into account when computing"
+                " the metric value."
+            )
+        values = y_true
+        if not isinstance(values, tf.Tensor):
+            values = tf.convert_to_tensor(values, dtype=self.dtype)
+        elif values.dtype != self.dtype:
+            values = tf.cast(values, dtype=self.dtype)
```
Comment (WindQAQ) [inline tensorflow_addons/metrics/geometric_mean.py:72]: `tf.cast` has already done the check. Please use `values = tf.cast(y_true, dtype=self.dtype)` directly.

https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/ops/math_ops.py#L732-L792

File: tensorflow_addons/metrics/geometric_mean.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                log_v = tf.reduce_sum(log_v)
+            self.accure.assign_add(log_v)
+
+    def result(self) -> tf.Tensor:
+        if tf.math.is_inf(self.accure):
+            return tf.constant(0, dtype=self.dtype)
+        ret = tf.math.exp(self.accure / self.count)
+        print(self.accure, self.count)
+        if ret.dtype is not self.dtype:
+            return tf.cast(ret, dtype=self.dtype)
```
Comment (WindQAQ) [inline tensorflow_addons/metrics/geometric_mean.py:87]: ditto

File: tensorflow_addons/metrics/geometric_mean.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    ):
+        super().__init__(name=name, dtype=dtype, **kwargs)
+        self.accure = self.add_weight(
+            "accure", shape=None, initializer="zeros", dtype=dtype
+        )
+        self.count = self.add_weight(
+            "count", shape=None, initializer="zeros", dtype=dtype
+        )
+
+    def update_state(self, y_true, y_pred=None, sample_weight=None) -> None:
```
Comment (WindQAQ) [inline tensorflow_addons/metrics/geometric_mean.py:59]: Maybe we can follow the input signature of `tf.keras.metrics.Mean()` like `def update_state(self, values, sample_weight=None)`. I'm aware our metrics contribution template does not take this into account.

cc @AakashKumarNain 

https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/keras/metrics.py#L314

File: tensorflow_addons/metrics/geometric_mean.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        if not isinstance(values, tf.Tensor):
+            values = tf.convert_to_tensor(values, dtype=self.dtype)
+        elif values.dtype != self.dtype:
+            values = tf.cast(values, dtype=self.dtype)
+
+        self.count.assign_add(np.size(values))
+        if not tf.math.is_inf(self.accure):
+            log_v = tf.math.log(values)
+            if log_v.shape != []:
+                log_v = tf.reduce_sum(log_v)
```
Comment (WindQAQ) [inline tensorflow_addons/metrics/geometric_mean.py:78]: just use `tf.reduce_sum(log_v)` without the condition, this makes it more clearly.

File: tensorflow_addons/metrics/geometric_mean.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            self.accure.assign_add(log_v)
+
+    def result(self) -> tf.Tensor:
+        if tf.math.is_inf(self.accure):
+            return tf.constant(0, dtype=self.dtype)
+        ret = tf.math.exp(self.accure / self.count)
+        print(self.accure, self.count)
```
Comment (WindQAQ) [inline tensorflow_addons/metrics/geometric_mean.py:85]: remove this debug line

File: tensorflow_addons/metrics/geometric_mean.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class GeometricMean(Metric):
+    """Compute Geometric Mean
+
+    The geometric mean is a kind of mean. Unlike the arithmetic mean
+    that uses the sum of values, it uses the product of the values to
+    represent typical values in a set of numbers.
+
+    Note: `tfa.metrics.GeometryMean` can be used the same as `tf.keras.metrics.Mean`
```
Comment (WindQAQ) [inline tensorflow_addons/metrics/geometric_mean.py:36]: nit: `tfa.metrics.GeometricMean`

File: tensorflow_addons/metrics/tests/geometric_mean_test.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        ([0, 0, 0, 0, 0, 0, 0, 1, 2, 6], 0),
+        ([0.2, 0.5, 0.3, 0.6, 0.1, 0.7], 0.32864603),
+        ([8, 4, 1, 7, 2, 11, 9, 22, 52], 7.1804023),
+        ([8.2, 9.7, 9.1, 2.7, 1.1, 2.0], 4.0324492),
+        ([0.6666666, 0.215213, 0.15167], 0.27918512),
+    ]
+
+
+def assert_result(expected, result):
+    np.testing.assert_allclose(expected, result, atol=1e-6)
```
Comment (WindQAQ) [inline tensorflow_addons/metrics/tests/geometric_mean_test.py:36]: this seems just a one line wrapper of `np.testing.assert_allclose`, we could probably use it directly.

File: tensorflow_addons/metrics/geometric_mean.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Comment (WindQAQ) [inline tensorflow_addons/metrics/geometric_mean.py:1]: 2020

File: tensorflow_addons/metrics/tests/geometric_mean_test.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Comment (WindQAQ) [inline tensorflow_addons/metrics/tests/geometric_mean_test.py:1]: ditto

File: tensorflow_addons/metrics/geometric_mean.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+            "accure", shape=None, initializer="zeros", dtype=dtype
+        )
+        self.count = self.add_weight(
+            "count", shape=None, initializer="zeros", dtype=dtype
+        )
+
+    def update_state(self, y_true, y_pred=None, sample_weight=None) -> None:
+        if y_pred is not None:
+            warnings.warn("`y_pred` is not None. `y_pred` is not used in GeometricMean")
+        if sample_weight is not None:
```
Comment (bhack) [inline tensorflow_addons/metrics/geometric_mean.py:62]: The only case across `Tesorflow` and `TFA` with this behavior is https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/multilabel_confusion_matrix.py#L111

Generally when it is computable it is expected to be computed (in the case we still want to maintain an homogeneous expectation):
 https://en.wikipedia.org/wiki/Weighted_geometric_mean

File: tensorflow_addons/metrics/geometric_mean.py (hunk 3 above)
Comment (marload) [reply tensorflow_addons/metrics/geometric_mean.py:59]: At first, I tried to follow the input signature, but I failed the test below and got the current signature.
Thank You!
https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/tests/metrics_test.py

File: tensorflow_addons/metrics/tests/geometric_mean_test.py (hunk 7 above)
Comment (marload) [reply tensorflow_addons/metrics/tests/geometric_mean_test.py:36]: This is a wrapper who uses it to fix `atol`. Personally, I think this is better.

File: tensorflow_addons/metrics/geometric_mean.py (hunk 3 above)
Comment (WindQAQ) [reply tensorflow_addons/metrics/geometric_mean.py:59]: Got it. You can also modify that file to make it pass the tests.

File: tensorflow_addons/metrics/tests/geometric_mean_test.py (hunk 7 above)
Comment (WindQAQ) [reply tensorflow_addons/metrics/tests/geometric_mean_test.py:36]: ok, then do what you want as this doesn't matter at all 😃

Comment (marload) [reply tensorflow_addons/metrics/tests/geometric_mean_test.py:36]: Thank You! 😄

**Comment by marload** [comment]
@bhack @WindQAQ 

I applied everything you mentioned!

Thank You!
